subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/ToolingApiVersionHelper.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal;
+
+import org.gradle.api.UncheckedIOException;
+import org.gradle.tooling.UnsupportedVersionException;
+import org.gradle.util.GradleVersion;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+public class ToolingApiVersionHelper {
```

Comment:
❌ Instead of creating a "Helper", try finding the right home for this code. It probably belongs in a deprecation-logging wrapper around `ConsumerConnection`. 

Also, in the tooling implementation loader, we already know which version is deprecated and which one is unsupported, so we can have an "UnsupportedConnection" and a "DeprecatedConnection", which don't need to do any further checking, but only throw/log on every request. This will get rid of a few if-statements.

Reply:
Thanks for your brilliant advice. A decorator is exactly what we need here.